### ISSUE TITLE
DTensor slicing on sharded dimension leads to replication

### ISSUE THREAD

**Issue by garrett361** [pytorch/pytorch#149447]
Slicing of sharded `DTensor`s currently results in differing placements depending on the axis over which the `DTensor` is sharded:
* Slicing a sharded dimension leads to replication over that dimension.
* Slicing a replicated dimension preserves all placements of the `DTensor`

**Expectation**: slicing will preserve the placements of the original `DTensor`.  

Sketch of the behavior here:
```py
# Create sharded DTensor
x_dt = distribute_tensor(torch.randn(32, 32), mesh, (Shard(-1), ))

# Slice along the unsharded, zero dim. Slice is still sharded:
x_dt_slice0 = x_dt[:16]
x_dt_slice0.placement = (Shard(-1),) 

# Slice along the sharded -1 dim. Slice is now replicated
x_dt_slice0 = x_dt[..., :16]
x_dt_slice0.placement = (Replicate(),) 
```

[Minimal gist repro here.](https://gist.github.com/garrett361/f36b6c0b673cb1d777cb92f35438648c)

The asymmetric behavior seems undesirable, and seems liable to cause large memory spikes, e.g. if slicing a context-parallel sharded tensor.  Preserving the sharding placement with `torch.chunk` semantics seems like a natural expectation, IMO.

More generally, I think a reasonable expectation is for `distribute_tensor` and torch ops to commute:
```py
# Expectations

x = torch.rand(256)

# Case 1) distribute then slice
x_dt_slice0 = distribute_tensor(x, mesh, (Shard(-1),))[64:128]

# Case 2): slice then distribute
x_dt_slice1 = distribute_tensor(x[64:128], mesh, (Shard(-1),))

# These currently fail
assert x_dt_slice0.to_local().numel() == x_dt_slice1.to_local().numel()
assert x_dt_slice0.placements == x_dt_slice1.placements
```

**Comment by XilunWu** [comment]
> Slicing a sharded dimension leads to replication over that dimension.

I feel this is hard if not impossible to support. Take the following example:
```
mesh = init_device_mesh([4,])  # 4 ranks
dt = distribute_tensor(torch.ones(4, 4), mesh=mesh, placements=[Shard(0)])
dt_slice = dt[3]
```

It's hard to define the result DTensor without casting it to `Replicate`.

**Comment by wconstab** [comment]
Re-reading this issue: i think there are 2 factors

1) if we slice a sharded dimension, especially if it happens to be unevenly sharded, we may need to perform communication to get the right local values for the new dtensor.

Should we allow implicit communication to occur during a slice, or do users expect a slice (like a view) to be a "fast local operation"?  I think  

2) depending on (1) we can address the consistency problem in 2 ways
(a) if we make slice+shard raise an error when it requires communication, it would force the user to explicitly redistribute to replicate before slicing, and then they could choose to re-shard the result.
(b) if we do communication implicitly as part of the slice, we can replicate as an implementation detail for the slicing, and then we can re-shard to the orignial placement.

2(b) obviously still has the caveat that if we sliced a sharded dimension away, we would have to drop the sharding placement altogether.  I think this is OK though?

**Comment by garrett361** [comment]
While I generally favor of explicit over implicit, wouldn't the below force users to rewrite code to use DTensor in these cases?

> if we make slice+shard raise an error when it requires communication, it would force the user to explicitly redistribute to replicate before slicing, and then they could choose to re-shard the result.

That would seem to go against the nice, just-works goal of DTensor, imo.

> if we do communication implicitly as part of the slice, we can replicate as an implementation detail for the slicing, and then we can re-shard to the orignial placement.

The above is basically what I expected, but with intermediate replication avoided where possible.  Could easily take a lot of work to implement well, though.

**Comment by garrett361** [comment]
And regarding this:

> Should we allow implicit communication to occur during a slice, or do users expect a slice (like a view) to be a "fast local operation"?

I think users would typically expect the latter, but it doesn't seem possible? The current behavior also already has implicit comms.

**Comment by m3rcuriel** [comment]
It was surprising to me that the output of `nn.Linear` with `Replicate()` weights and `Shard(1)` input ends up producing a `Replicate()` output due to the `aten.view` op.

```
Input: torch.Size([2, 512, 4096]) (Shard(dim=1),)
Weight: torch.Size([14336, 4096]) (Replicate(),)
Dispatching aten.t.default
Input placements: [(Replicate(),)]
Output placement: (Replicate(),)
Dispatching aten.view.default
Input placements: [(Shard(dim=1),), 'not DTensor']
Output placement: (Replicate(),)
Dispatching aten.mm.default
Input placements: [(Replicate(),), (Replicate(),)]
Output placement: (Replicate(),)
Dispatching aten._unsafe_view.default
Input placements: [(Replicate(),), 'not DTensor']
Output placement: (Replicate(),)
Output: torch.Size([2, 512, 14336]) (Replicate(),)
```
